### ISSUE TITLE
Scope for Secp256r1 curve?

### ISSUE THREAD

**Issue by Nilay27** [personaelabs/spartan-ecdsa#36]
While this library makes signature verification of secp256k1-based curves much more efficient, could this also be adapted to secp256r1 curves (P-256)? 

The reason being P-256 is among the most common signature scheme that laptops and mobiles support (including secure enclave for Apple devices). 
However, the signature verification of P-256-based curves costs a lot of gas (around 1.2 million) when they are verified on smart contracts. 

A significant improvement could be made if we are able to verify the P-256-based signatures inside ZK and then present the proof to the Smart contract. Using Snarks, this would cost roughly 250k gas which is a huge reduction from the previous 1.2 million gas and would be equally secure. 

I saw that the `Secp256k1AddIncomplete` and `Secp256k1Double` work for any Weierstrass Curve, does this mean `Secp256k1Mul` and `Secp256k1AddComplete` would also work for any Weierstrass curve if we just change the parameters of the `template k()`?

**Comment by DanTehrani** [comment]
@Nilay27 
P-256 would be an interesting curve to support, and I believe you only need to change some parameters to use the circuits of P256 signature verification.
- I saw that the `Secp256k1AddIncomplete` and `Secp256k1Double` work for any Weierstrass Curve, does this mean `Secp256k1Mul` and `Secp256k1AddComplete` would also work for any Weierstrass curve if we just change the parameters of the `template k()`?

But the issue is that spartan-ecdsa proofs can’t be verified on-chain (the proof size is > 10KB) and the verification consists of many scalar multiplications.

In other words, spartan-ecdsa is not meant to “compress” statements as SNARKs do, and it only considers proving statements in zero knowledge.

I'm currently [working on an FRI-based proof](https://github.com/DanTehrani/spartan-fri) system that supports proving on fields of existing curves (P256, Secp256k1) and also that can support on-chain verification. So that could be a solution for your use case when it's ready!

**Comment by Nilay27** [comment]
> I'm currently [working on an FRI-based proof](https://github.com/DanTehrani/spartan-fri) system that supports proving on fields of existing curves (P256, Secp256k1) and also that can support on-chain verification. So that could be a solution for your use case when it's ready!
That would be great! looking forward to it.
